# Pandas

In [1]:
# semua package yang digunakan di file ini import disini
import numpy as np
import pandas as pd

In [2]:
df_nba = pd.read_csv('dataset/nbaallelo.csv')
print(type(df_nba))

<class 'pandas.core.frame.DataFrame'>


In [3]:
# bisa juga import dari link
# untuk assignment, pastikan file data yang kalian gunakan itu bisa di akses

# df = pd.read_csv('https://raw.githubusercontent.com/ardhiraka/PFDS_sources/master/nbaallelo.csv')
# df

## Melihat detail data frame
- shape (bentuk)
- head (baris baris pertama)
- tail (baris baris terakhir)
- info (informasi termasuk shape, tipe data, judul kolom, dll)
- describe (informasi yang berhubungan dengan statistik, angka2nya)

In [4]:
# lihat properti dari data yang sudah kita import, si nbaallelo.csv
print(df_nba.shape) # jumlah baris dan kolom
print(len(df_nba)) # jumlah baris

(126314, 23)
126314


shape menunjukkan ada 126314 baris dan 34 kolom di data nbaallelo.csv ini

In [5]:
# menunjukkan 5 baris pertama di dataframe
display(df_nba.head())

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...  40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...  41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...  42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...  40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...  38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

In [6]:
# bisa juga custom mau tunjukkan berapa baris pertama
display(df_nba.head(2))

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...   40.29483     NYK    Knicks       68   
1            0     NYK   Knicks  ...   41.70517     TRH   Huskies       66   

  opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0    1300.0  1306.7233              H            L  0.640065   NaN  
1    1300.0  1293.2767              A            W  0.359935   NaN  

[2 rows x 23 columns]

In [7]:
# bisa juga melihat baris baris terakhir
display(df_nba.tail(2))

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
126312         102            1     CLE  Cavaliers  ...  59.290245     GSW   
126313         103            1     GSW   Warriors  ...  68.519516     CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  \
126312   Warriors      105  1813.6349  1822.2881              H            L   
126313  Cavaliers       97  1700.7391  1692.0859              A            W   

        forecast notes  
126312   0.48145   NaN  
126313   0.51855   NaN  

[2 rows x 23 columns]

In [8]:
display(df_nba.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   gameorder      126314 non-null  int64  
 1   game_id        126314 non-null  object 
 2   lg_id          126314 non-null  object 
 3   _iscopy        126314 non-null  int64  
 4   year_id        126314 non-null  int64  
 5   date_game      126314 non-null  object 
 6   seasongame     126314 non-null  int64  
 7   is_playoffs    126314 non-null  int64  
 8   team_id        126314 non-null  object 
 9   fran_id        126314 non-null  object 
 10  pts            126314 non-null  int64  
 11  elo_i          126314 non-null  float64
 12  elo_n          126314 non-null  float64
 13  win_equiv      126314 non-null  float64
 14  opp_id         126314 non-null  object 
 15  opp_fran       126314 non-null  object 
 16  opp_pts        126314 non-null  int64  
 17  opp_elo_i      126314 non-nul

None

- null - kosong
- NaN - not a number, bukan angka, termasuk "kosong"
- None - kosong

In [9]:
display(df_nba.describe())

gameorder        _iscopy        year_id     seasongame  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean    31579.000000       0.500000    1988.200374      43.533733   
std     18231.927643       0.500002      17.582309      25.375178   
min         1.000000       0.000000    1947.000000       1.000000   
25%     15790.000000       0.000000    1975.000000      22.000000   
50%     31579.000000       0.500000    1990.000000      43.000000   
75%     47368.000000       1.000000    2003.000000      65.000000   
max     63157.000000       1.000000    2015.000000     108.000000   

         is_playoffs            pts          elo_i          elo_n  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean        0.063857     102.729982    1495.236055    1495.236055   
std         0.244499      14.814845     112.139945     112.461687   
min         0.000000       0.000000    1091.644500    1085.774400   
25%         0.000000      93.000000    1417.237975    1416.994900   
50%         0.000000     103.000000    1500.945550    1500.954400   
75%         0.000000     112.000000    1576.060000    1576.291625   
max         1.000000     186.000000    1853.104500    1853.104500   

           win_equiv        opp_pts      opp_elo_i      opp_elo_n  \
count  126314.000000  126314.000000  126314.000000  126314.000000   
mean       41.707889     102.729982    1495.236055    1495.236055   
std        10.627332      14.814845     112.139945     112.461687   
min        10.152501       0.000000    1091.644500    1085.774400   
25%        34.103035      93.000000    1417.237975    1416.994900   
50%        42.113357     103.000000    1500.945550    1500.954400   
75%        49.635328     112.000000    1576.060000    1576.291625   
max        71.112038     186.000000    1853.104500    1853.104500   

            forecast  
count  126314.000000  
mean        0.500000  
std         0.215252  
min         0.020447  
25%         0.327989  
50%         0.500000  
75%         0.672011  
max         0.979553

### value_counts()
untuk melihat ada value apa saja di sebuah kolom, dan setiap value muncul berapa kali

In [12]:
df_nba['lg_id'].value_counts()

NBA    118016
ABA      8298
Name: lg_id, dtype: int64

### nunique()
mirip value counts tapi tidak memperlihatkan value nya apa saja, dan total muncul berapa kali. hanya memperlihatkan kolom ini punya berapa unique values.

In [15]:
df_nba['lg_id'].nunique()

2

In [16]:
df_nba['gameorder'].nunique()

63157

In [17]:
63157*2

126314

### pd.to_datetime()
mengubah tipe data menjadi datetime untuk tanggalan

In [13]:
df_nba['date_game'] = pd.to_datetime(df_nba['date_game'])

In [14]:
display(df_nba.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   gameorder      126314 non-null  int64         
 1   game_id        126314 non-null  object        
 2   lg_id          126314 non-null  object        
 3   _iscopy        126314 non-null  int64         
 4   year_id        126314 non-null  int64         
 5   date_game      126314 non-null  datetime64[ns]
 6   seasongame     126314 non-null  int64         
 7   is_playoffs    126314 non-null  int64         
 8   team_id        126314 non-null  object        
 9   fran_id        126314 non-null  object        
 10  pts            126314 non-null  int64         
 11  elo_i          126314 non-null  float64       
 12  elo_n          126314 non-null  float64       
 13  win_equiv      126314 non-null  float64       
 14  opp_id         126314 non-null  object        
 15  

None

## Explore Dataset
kita akan eksplorasi dataset kita, untuk lebih mengenal trend datanya, dan apa yg terjadi (mencari informasi)

In [19]:
# misal, tim mana yang paling sering main
# ada berapa tim?

df_nba['team_id'].value_counts()

BOS    5997
NYK    5769
LAL    5078
DET    4985
PHI    4533
       ... 
DTF      60
INJ      60
PIT      60
TRH      60
SDS      11
Name: team_id, Length: 104, dtype: int64

In [20]:
# fran_id nama franchise team nya
# apakah ada tim yang nama franchise nya sama?
df_nba['fran_id'].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Denver       

dari jumlah value_counts team_id dan fran_id, terlihat kalau ada lebih dari 1 tim yang namanya sama (tapi id unik untuk masing2 tim)

In [22]:
# kita cek tim Lakers dari mana saja, selain LAL ada siapa lagi

tim_lakers = df_nba['fran_id'] == 'Lakers'
df_nba[tim_lakers]

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
1136          569  194811040BLB   NBA        1     1949 1948-11-04   
1152          577  194811060WSC   NBA        1     1949 1948-11-06   
1159          580  194811090BOS   NBA        1     1949 1948-11-09   
1170          586  194811110NYK   NBA        1     1949 1948-11-11   
1183          592  194811130PRO   NBA        1     1949 1948-11-13   
...           ...           ...   ...      ...      ...        ...   
126016      63009  201504080DEN   NBA        1     2015 2015-04-08   
126052      63027  201504100LAL   NBA        0     2015 2015-04-10   
126086      63044  201504120LAL   NBA        0     2015 2015-04-12   
126115      63058  201504130SAC   NBA        1     2015 2015-04-13   
126137      63069  201504150LAL   NBA        0     2015 2015-04-15   

        seasongame  is_playoffs team_id fran_id  ...  win_equiv  opp_id  \
1136             1            0     MNL  Lakers  ...  51.218754     BLB   
1152             2            0     MNL  Lakers  ...  50.621624     WSC   
1159             3            0     MNL  Lakers  ...  48.341133     BOS   
1170             4            0     MNL  Lakers  ...  49.499451     NYK   
1183             5            0     MNL  Lakers  ...  50.330185     PRO   
...            ...          ...     ...     ...  ...        ...     ...   
126016          78            0     LAL  Lakers  ...  21.921890     DEN   
126052          79            0     LAL  Lakers  ...  22.371969     MIN   
126086          80            0     LAL  Lakers  ...  21.791288     DAL   
126115          81            0     LAL  Lakers  ...  21.422607     SAC   
126137          82            0     LAL  Lakers  ...  19.985550     SAC   

            opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  \
1136       Baltimore       72  1507.2052  1491.9507              A   
1152        Capitols       67  1481.8721  1487.9188              A   
1159         Celtics       77  1359.3345  1382.0542              A   
1170          Knicks       68  1460.4232  1448.9517              A   
1183    Steamrollers       67  1245.9456  1237.6326              A   
...              ...      ...        ...        ...            ...   
126016       Nuggets      119  1415.0565  1421.3248              A   
126052  Timberwolves       98  1283.9249  1278.4835              H   
126086     Mavericks      120  1556.2162  1563.2498              H   
126115         Kings      102  1395.6000  1400.1278              A   
126137         Kings      122  1400.1278  1418.2802              H   

        game_result  forecast notes  
1136              W  0.386583   NaN  
1152              L  0.443232   NaN  
1159              L  0.608859   NaN  
1170              W  0.432863   NaN  
1183              W  0.737010   NaN  
...             ...       ...   ...  
126016            L  0.238914   NaN  
126052            W  0.670716   NaN  
126086            L  0.304769   NaN  
126115            L  0.251263   NaN  
126137            L  0.501819   NaN  

[6024 rows x 23 columns]

In [26]:
df_lakers = df_nba[df_nba['fran_id'] == 'Lakers']
display(df_lakers.shape)

(6024, 23)

In [27]:
# team_id nya kita value_counts
df_lakers['team_id'].value_counts()

LAL    5078
MNL     946
Name: team_id, dtype: int64

kita menemukan adanya tim Minneapolis Lakers selain LA Lakers

In [30]:
# Minneapolis Lakers game nya ada dari kapan sampai kapan? cari tanggal game pertama dan game terakhir tim MNL

df_mnl = df_nba[df_nba['team_id'] == 'MNL']
print(df_mnl.shape)
display(df_mnl)

(946, 23)


gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
1136        569  194811040BLB   NBA        1     1949 1948-11-04           1   
1152        577  194811060WSC   NBA        1     1949 1948-11-06           2   
1159        580  194811090BOS   NBA        1     1949 1948-11-09           3   
1170        586  194811110NYK   NBA        1     1949 1948-11-11           4   
1183        592  194811130PRO   NBA        1     1949 1948-11-13           5   
...         ...           ...   ...      ...      ...        ...         ...   
9701       4851  196003190LAL   NBA        0     1960 1960-03-19          80   
9702       4852  196003200LAL   NBA        0     1960 1960-03-20          81   
9709       4855  196003220STL   NBA        1     1960 1960-03-22          82   
9711       4856  196003240LAL   NBA        0     1960 1960-03-24          83   
9715       4858  196003260STL   NBA        1     1960 1960-03-26          84   

      is_playoffs team_id fran_id  ...  win_equiv  opp_id      opp_fran  \
1136            0     MNL  Lakers  ...  51.218754     BLB     Baltimore   
1152            0     MNL  Lakers  ...  50.621624     WSC      Capitols   
1159            0     MNL  Lakers  ...  48.341133     BOS       Celtics   
1170            0     MNL  Lakers  ...  49.499451     NYK        Knicks   
1183            0     MNL  Lakers  ...  50.330185     PRO  Steamrollers   
...           ...     ...     ...  ...        ...     ...           ...   
9701            1     MNL  Lakers  ...  36.052319     STL         Hawks   
9702            1     MNL  Lakers  ...  36.565147     STL         Hawks   
9709            1     MNL  Lakers  ...  38.137211     STL         Hawks   
9711            1     MNL  Lakers  ...  36.154404     STL         Hawks   
9715            1     MNL  Lakers  ...  35.694744     STL         Hawks   

      opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  forecast  \
1136       72  1507.2052  1491.9507              A            W  0.386583   
1152       67  1481.8721  1487.9188              A            L  0.443232   
1159       77  1359.3345  1382.0542              A            L  0.608859   
1170       68  1460.4232  1448.9517              A            W  0.432863   
1183       67  1245.9456  1237.6326              A            W  0.737010   
...       ...        ...        ...            ...          ...       ...   
9701       93  1564.9653  1571.4325              H            L  0.508790   
9702      101  1571.4325  1566.4786              H            W  0.490178   
9709      110  1566.4786  1551.3701              A            W  0.243506   
9711      117  1551.3701  1570.4452              H            L  0.547775   
9715       97  1570.4452  1574.8955              A            L  0.235193   

     notes  
1136   NaN  
1152   NaN  
1159   NaN  
1170   NaN  
1183   NaN  
...    ...  
9701   NaN  
9702   NaN  
9709   NaN  
9711   NaN  
9715   NaN  

[946 rows x 23 columns]

In [31]:
df_mnl['date_game'].min()

Timestamp('1948-11-04 00:00:00')

In [32]:
df_mnl['date_game'].max()

Timestamp('1960-03-26 00:00:00')

In [33]:
df_nba.loc[df_nba['team_id'] == 'MNL', 'date_game']

1136   1948-11-04
1152   1948-11-06
1159   1948-11-09
1170   1948-11-11
1183   1948-11-13
          ...    
9701   1960-03-19
9702   1960-03-20
9709   1960-03-22
9711   1960-03-24
9715   1960-03-26
Name: date_game, Length: 946, dtype: datetime64[ns]

# Mengenal Pandas Series dan Pandas DataFrame

## Series
1 dimensi

In [34]:
revenues = pd.Series([5000,7000,9000])
print(revenues)
print(type(revenues))

0    5000
1    7000
2    9000
dtype: int64
<class 'pandas.core.series.Series'>


In [35]:
revenues = pd.Series([5000,7000,9000], index=["Senin", "Selasa", "Rabu"])
print(revenues)
print(type(revenues))

Senin     5000
Selasa    7000
Rabu      9000
dtype: int64
<class 'pandas.core.series.Series'>


In [39]:
revenues = pd.Series({"Senin":5000, "Selasa":7000, "Rabu":9000})
print(revenues)
print(type(revenues))

Senin     5000
Selasa    7000
Rabu      9000
dtype: int64
<class 'pandas.core.series.Series'>


In [40]:
# revenues = pd.Series({"Senin":[5000], "Selasa":7000, "Rabu":9000})
# print(revenues)
# print(type(revenues))

In [49]:
# mirip dictionary kita bisa cek values nya dan index nya apa saja
print(revenues.values)
print(revenues.index)
print(revenues.keys()) # keys di series adalah index nya
print(revenues.shape)

[5000 7000 9000]
Index(['Senin', 'Selasa', 'Rabu'], dtype='object')
Index(['Senin', 'Selasa', 'Rabu'], dtype='object')
(3,)


In [48]:
# series tidak punya info, masih punya describe
print(revenues.describe())

count       3.0
mean     7000.0
std      2000.0
min      5000.0
25%      6000.0
50%      7000.0
75%      8000.0
max      9000.0
dtype: float64


In [50]:
# cara edit data di series
revenues['Selasa'] = 8000

print(revenues)

Senin     5000
Selasa    8000
Rabu      9000
dtype: int64


In [74]:
# kalau mau ubah index Series lebih rumit sedikit, kita perlu reset_index dulu, lalu reindex
rev2 = revenues.reset_index()
display(rev2)
print(type(rev2))

index     0
0   Senin  5000
1  Selasa  8000
2    Rabu  9000

<class 'pandas.core.frame.DataFrame'>


In [72]:
rev2.loc[1, 'index'] = 'Kamis'
print(rev2)

   index     0
0  Senin  5000
1  Kamis  8000
2   Rabu  9000


In [65]:
rev2[0]

0    5000
1    8000
2    9000
Name: 0, dtype: int64

In [73]:
new_revenue = pd.Series(data=np.array(rev2[0]), index=np.array(rev2['index']))
display(new_revenue)

Senin    5000
Kamis    8000
Rabu     9000
dtype: int64

In [78]:
# balik ke contoh awal kita series revenues
display(revenues)
cab1 = revenues

Senin     5000
Selasa    8000
Rabu      9000
dtype: int64

In [79]:
# kita buat series baru
penjualan = pd.Series({"Senin":8000, "Rabu":3500, "Kamis":6700})
display(penjualan)
cab2 = penjualan

Senin    8000
Rabu     3500
Kamis    6700
dtype: int64

## DataFrame
- cara buat, bisa dari series, bisa dari list juga, np array juga bisa
- bisa import data dari luar read_csv, read_excel, ...
- bisa lihat len, shape, info, describe, head, tail
- sama seperti Series, bisa lihat index, values, keys, columns

In [80]:
# kita bisa gabungkan beberapa Series jadi satu DataFrame
df_penjualan = pd.DataFrame({"Cabang1": cab1, "Cabang2": cab2})
display(df_penjualan)

Cabang1  Cabang2
Kamis       NaN   6700.0
Rabu     9000.0   3500.0
Selasa   8000.0      NaN
Senin    5000.0   8000.0

In [81]:
df_penjualan.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, Kamis to Senin
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Cabang1  3 non-null      float64
 1   Cabang2  3 non-null      float64
dtypes: float64(2)
memory usage: 96.0+ bytes


In [82]:
df_penjualan.describe()

Cabang1      Cabang2
count     3.000000     3.000000
mean   7333.333333  6066.666667
std    2081.665999  2315.887159
min    5000.000000  3500.000000
25%    6500.000000  5100.000000
50%    8000.000000  6700.000000
75%    8500.000000  7350.000000
max    9000.000000  8000.000000

In [83]:
df_penjualan.index

Index(['Kamis', 'Rabu', 'Selasa', 'Senin'], dtype='object')

In [84]:
df_penjualan.values

array([[  nan, 6700.],
       [9000., 3500.],
       [8000.,   nan],
       [5000., 8000.]])

In [85]:
df_penjualan.keys() # kalau di DataFrame, keys adalah judul kolom

Index(['Cabang1', 'Cabang2'], dtype='object')

In [86]:
df_penjualan.columns

Index(['Cabang1', 'Cabang2'], dtype='object')

In [87]:
df_penjualan.axes # melihat index dan columns sebuah dataframe

[Index(['Kamis', 'Rabu', 'Selasa', 'Senin'], dtype='object'),
 Index(['Cabang1', 'Cabang2'], dtype='object')]

# operator in

In [88]:
display(revenues)

Senin     5000
Selasa    8000
Rabu      9000
dtype: int64

In [90]:
"Senin" in revenues # nyari dari .keys(), Series -> index

True

In [93]:
"Selasa" in revenues.index

True

In [91]:
display(df_penjualan)

Cabang1  Cabang2
Kamis       NaN   6700.0
Rabu     9000.0   3500.0
Selasa   8000.0      NaN
Senin    5000.0   8000.0

In [92]:
"Cabang1" in df_penjualan # nyari dari .keys(), DataFrame -> columns

True

In [94]:
"Cabang1" in df_penjualan.columns

True

In [95]:
"Senin" in df_penjualan.index

True

## loc dan iloc
- loc -> mencari dari index (text index)
- iloc -> dari index angka, baris ke berapa

In [96]:
display(revenues)

Senin     5000
Selasa    8000
Rabu      9000
dtype: int64

In [97]:
revenues.loc['Selasa']

8000

In [100]:
revenues.iloc[1]

8000

In [99]:
revenues.iloc[-1]

9000

In [105]:
revenues.loc['Senin':'Selasa']

Senin     5000
Selasa    8000
dtype: int64

In [107]:
revenues.iloc[0:2]

Senin     5000
Selasa    8000
dtype: int64

In [101]:
display(df_penjualan)

Cabang1  Cabang2
Kamis       NaN   6700.0
Rabu     9000.0   3500.0
Selasa   8000.0      NaN
Senin    5000.0   8000.0

In [102]:
df_penjualan.loc['Rabu']

Cabang1    9000.0
Cabang2    3500.0
Name: Rabu, dtype: float64

In [103]:
df_penjualan.iloc[1]

Cabang1    9000.0
Cabang2    3500.0
Name: Rabu, dtype: float64

In [104]:
df_penjualan.iloc[-1]

Cabang1    5000.0
Cabang2    8000.0
Name: Senin, dtype: float64

In [109]:
df_penjualan.loc['Rabu':'Selasa']

Cabang1  Cabang2
Rabu     9000.0   3500.0
Selasa   8000.0      NaN

In [110]:
df_penjualan.iloc[1:3]

Cabang1  Cabang2
Rabu     9000.0   3500.0
Selasa   8000.0      NaN

In [111]:
# untuk dataframe, selain baris baris tertentu, kita bisa pilih untuk ambil kolom tertentu juga
df_penjualan.loc['Rabu':'Selasa', 'Cabang1']

Rabu      9000.0
Selasa    8000.0
Name: Cabang1, dtype: float64

In [112]:
df_penjualan.iloc[1:3, 0]

Rabu      9000.0
Selasa    8000.0
Name: Cabang1, dtype: float64

In [113]:
df_penjualan.loc['Rabu':'Selasa', 'Cabang1':'Cabang2']

Cabang1  Cabang2
Rabu     9000.0   3500.0
Selasa   8000.0      NaN

In [114]:
df_penjualan.iloc[1:3, 0:2]

Cabang1  Cabang2
Rabu     9000.0   3500.0
Selasa   8000.0      NaN

In [116]:
notes = pd.Series({'Senin':"diskon nomad", 'Selasa':"happy hour", 'Rabu':"ladies night", 'Kamis':"pre-weekend"})
df_baru = pd.DataFrame({'Cabang1':cab1, 'Cabang2':cab2, 'Notes':notes})
display(df_baru)

Cabang1  Cabang2         Notes
Kamis       NaN   6700.0   pre-weekend
Rabu     9000.0   3500.0  ladies night
Selasa   8000.0      NaN    happy hour
Senin    5000.0   8000.0  diskon nomad

In [117]:
df_baru.loc[['Senin', 'Selasa'], ['Cabang1', 'Notes']]

Cabang1         Notes
Senin    5000.0  diskon nomad
Selasa   8000.0    happy hour

In [118]:
df_baru.iloc[[-1, -2], [0, -1]]

Cabang1         Notes
Senin    5000.0  diskon nomad
Selasa   8000.0    happy hour

# Querying Dataset
kita bisa buat kondisi/ pertanyaan untuk mengambil data tertentu dari dataset kita

In [119]:
display(df_baru)

Cabang1  Cabang2         Notes
Kamis       NaN   6700.0   pre-weekend
Rabu     9000.0   3500.0  ladies night
Selasa   8000.0      NaN    happy hour
Senin    5000.0   8000.0  diskon nomad

In [121]:
# misal kita ambil yang datanya lebih dari 5000 di Cabang1
df_baru[df_baru['Cabang1'] > 5000]

Cabang1  Cabang2         Notes
Rabu     9000.0   3500.0  ladies night
Selasa   8000.0      NaN    happy hour

In [124]:
cabang1_5000plus = df_baru['Cabang1'] >= 5000
cabang2_5000plus = df_baru['Cabang2'] >= 5000

# cari penjualan yang cabang1 atau cabang2 lebih dari 5000
df_baru[cabang1_5000plus | cabang2_5000plus]

Cabang1  Cabang2         Notes
Kamis       NaN   6700.0   pre-weekend
Rabu     9000.0   3500.0  ladies night
Selasa   8000.0      NaN    happy hour
Senin    5000.0   8000.0  diskon nomad

In [125]:
# cari penjualan yang cabang1 dan cabang2 lebih dari 5000
df_baru[cabang1_5000plus & cabang2_5000plus]

Cabang1  Cabang2         Notes
Senin   5000.0   8000.0  diskon nomad

In [126]:
# untuk data yang kita mau, salah satu cabang penjualan 5000+ dan mau lihat kolom notes saja
df_baru[cabang1_5000plus & cabang2_5000plus]['Notes']

Senin    diskon nomad
Name: Notes, dtype: object

In [127]:
df_baru.loc[cabang1_5000plus & cabang2_5000plus, 'Notes']

Senin    diskon nomad
Name: Notes, dtype: object

In [128]:
df_baru.loc[cabang1_5000plus & cabang2_5000plus, ['Notes']]

Notes
Senin  diskon nomad

## .notnull()
mencari data yang tidak null, termasuk querying dataset tapi ini yang built-in

In [129]:
display(df_baru)

Cabang1  Cabang2         Notes
Kamis       NaN   6700.0   pre-weekend
Rabu     9000.0   3500.0  ladies night
Selasa   8000.0      NaN    happy hour
Senin    5000.0   8000.0  diskon nomad

In [132]:
df_baru.notnull()

Cabang1  Cabang2  Notes
Kamis     False     True   True
Rabu       True     True   True
Selasa     True    False   True
Senin      True     True   True

In [133]:
df_baru['Cabang1'].notnull()

Kamis     False
Rabu       True
Selasa     True
Senin      True
Name: Cabang1, dtype: bool

In [134]:
df_baru[df_baru['Cabang1'].notnull()]

Cabang1  Cabang2         Notes
Rabu     9000.0   3500.0  ladies night
Selasa   8000.0      NaN    happy hour
Senin    5000.0   8000.0  diskon nomad

In [135]:
# bisa kita gabung juga dengan and & dan or |
df_baru[df_baru['Cabang1'].notnull() & df_baru['Cabang2'].notnull()]

Cabang1  Cabang2         Notes
Rabu    9000.0   3500.0  ladies night
Senin   5000.0   8000.0  diskon nomad

In [136]:
# contoh berikutnya kita bisa cek juga sebagian dari value nya, misalkan data nba kita,
# khusus tim yang namanya dimulai dengan "S"

df_nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947 1946-11-01           1   
1          1  194611010TRH   NBA        1     1947 1946-11-01           1   
2          2  194611020CHS   NBA        0     1947 1946-11-02           1   
3          2  194611020CHS   NBA        1     1947 1946-11-02           2   
4          3  194611020DTF   NBA        0     1947 1946-11-02           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...  40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...  41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...  42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...  40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...  38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

In [140]:
df_nba[df_nba['fran_id'].str.startswith("S")]

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
2               2  194611020CHS   NBA        0     1947 1946-11-02   
7               4  194611020PRO   NBA        0     1947 1946-11-02   
15              8  194611050BOS   NBA        1     1947 1946-11-05   
20             11  194611070PRO   NBA        0     1947 1946-11-07   
21             11  194611070PRO   NBA        1     1947 1946-11-07   
...           ...           ...   ...      ...      ...        ...   
126192      63097  201504240SAS   NBA        0     2015 2015-04-24   
126209      63105  201504260SAS   NBA        0     2015 2015-04-26   
126221      63111  201504280LAC   NBA        1     2015 2015-04-28   
126228      63115  201504300SAS   NBA        0     2015 2015-04-30   
126233      63117  201505020LAC   NBA        1     2015 2015-05-02   

        seasongame  is_playoffs team_id       fran_id  ...  win_equiv  opp_id  \
2                1            0     CHS         Stags  ...  42.012257     NYK   
7                1            0     PRO  Steamrollers  ...  41.540619     BOS   
15               2            0     CHS         Stags  ...  42.686188     BOS   
20               2            0     PRO  Steamrollers  ...  42.206284     CHS   
21               3            0     CHS         Stags  ...  42.020947     PRO   
...            ...          ...     ...           ...  ...        ...     ...   
126192          85            1     SAS         Spurs  ...  63.336937     LAC   
126209          86            1     SAS         Spurs  ...  62.168911     LAC   
126221          87            1     SAS         Spurs  ...  62.804081     LAC   
126228          88            1     SAS         Spurs  ...  61.898045     LAC   
126233          89            1     SAS         Spurs  ...  61.634998     LAC   

            opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  \
2             Knicks       47  1306.7233  1297.0712              H   
7            Celtics       53  1300.0000  1294.8458              H   
15           Celtics       55  1294.8458  1288.4139              A   
20             Stags       65  1316.0840  1309.7350              H   
21      Steamrollers       73  1305.1542  1311.5032              A   
...              ...      ...        ...        ...            ...   
126192      Clippers       73  1706.8958  1694.7960              H   
126209      Clippers      114  1694.7960  1710.2974              H   
126221      Clippers      107  1710.2974  1701.9489              A   
126228      Clippers      102  1701.9489  1713.8020              H   
126233      Clippers      111  1713.8020  1717.1750              A   

        game_result  forecast notes  
2                 W  0.631101   NaN  
7                 W  0.640065   NaN  
15                W  0.379796   NaN  
20                W  0.625446   NaN  
21                L  0.374554   NaN  
...             ...       ...   ...  
126192            W  0.671681   NaN  
126209            L  0.701636   NaN  
126221            W  0.383515   NaN  
126228            L  0.684116   NaN  
126233            L  0.374021   NaN  

[17764 rows x 23 columns]

In [145]:
# kita bisa lanjutkan untuk cek tim nya apa saja pakai value_counts
df_nba.loc[df_nba['fran_id'].str.startswith("S"), 'fran_id'].value_counts()

Sixers          5644
Spurs           4309
Suns            4080
Squires          799
Spirits          777
Stars            756
Sounds           697
Sails            274
Stags            260
Steamrollers     168
Name: fran_id, dtype: int64

In [146]:
df_nba.loc[df_nba['fran_id'].str.startswith("S"), 'fran_id'].nunique()

10

In [150]:
df_nba.loc[df_nba['fran_id'].str.startswith("L"), 'fran_id'].nunique()

1

# Grouping and Aggregating data

In [153]:
display(df_penjualan)

Cabang1  Cabang2
Kamis       NaN   6700.0
Rabu     9000.0   3500.0
Selasa   8000.0      NaN
Senin    5000.0   8000.0

In [154]:
df_penjualan.sum()

Cabang1    22000.0
Cabang2    18200.0
dtype: float64

In [155]:
df_penjualan.max()

Cabang1    9000.0
Cabang2    8000.0
dtype: float64

In [156]:
df_penjualan.min()

Cabang1    5000.0
Cabang2    3500.0
dtype: float64

In [157]:
df_penjualan.mean()

Cabang1    7333.333333
Cabang2    6066.666667
dtype: float64

In [158]:
# bagaimana kalau kita mau cari rata2 skor setiap tim yang berbeda?
df_nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947 1946-11-01           1   
1          1  194611010TRH   NBA        1     1947 1946-11-01           1   
2          2  194611020CHS   NBA        0     1947 1946-11-02           1   
3          2  194611020CHS   NBA        1     1947 1946-11-02           2   
4          3  194611020DTF   NBA        0     1947 1946-11-02           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...  40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...  41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...  42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...  40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...  38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

In [159]:
df_nba.mean() # ini kan tidak bisa melihat per tim nya means nya berapa, jadi kita group dulu per tim

<ipython-input-159-dd467d8acbd2>:1: FutureWarning: DataFrame.mean and DataFrame.median with numeric_only=None will include datetime64 and datetime64tz columns in a future version.
  df_nba.mean()


gameorder      31579.000000
_iscopy            0.500000
year_id         1988.200374
seasongame        43.533733
is_playoffs        0.063857
pts              102.729982
elo_i           1495.236055
elo_n           1495.236055
win_equiv         41.707889
opp_pts          102.729982
opp_elo_i       1495.236055
opp_elo_n       1495.236055
forecast           0.500000
dtype: float64

In [161]:
df_nba.groupby('team_id').mean() # masih semua kolom yang ada

gameorder   _iscopy      year_id  seasongame  is_playoffs  \
team_id                                                                 
ANA       8031.730769  0.474359  1968.000000   39.500000     0.000000   
AND       1263.875000  0.472222  1950.000000   36.500000     0.111111   
ATL      34816.527881  0.496654  1991.977943   43.656753     0.056753   
BAL       9313.930716  0.438799  1968.593533   43.978060     0.061201   
BLB       1689.137045  0.404711  1951.111349   33.982869     0.036403   
...               ...       ...          ...         ...          ...   
WAS      51881.806102  0.500339  2006.711864   42.044746     0.031864   
WAT       1207.145161  0.500000  1950.000000   31.500000     0.000000   
WSA      10245.538462  0.626374  1970.000000   46.000000     0.076923   
WSB      27316.062751  0.500502  1985.676205   44.026606     0.053213   
WSC        798.862543  0.494845  1948.859107   31.295533     0.068729   

                pts        elo_i        elo_n  win_equiv     opp_pts  \
team_id                                                                
ANA      111.589744  1235.538082  1233.713450  34.197335  116.115385   
AND       86.013889  1524.701615  1524.192312  48.516213   82.555556   
ATL      101.920198  1491.203485  1491.162358  40.919264  102.259232   
BAL      113.293303  1456.299155  1456.289124  39.743730  113.711316   
BLB       79.698073  1391.456756  1390.778889  32.560422   83.349036   
...             ...          ...          ...        ...         ...   
WAS       96.551186  1439.937444  1439.699562  34.844581   98.915254   
WAT       79.370968  1314.718410  1313.881785  27.217966   84.903226   
WSA      118.395604  1440.020588  1438.910534  46.097352  119.736264   
WSB      104.214859  1480.436141  1480.376864  39.561319  105.227912   
WSC       76.931271  1465.601227  1465.692951  46.485884   74.965636   

           opp_elo_i    opp_elo_n  forecast  
team_id                                      
ANA      1303.618576  1305.443215  0.418746  
AND      1427.774661  1428.283972  0.624571  
ATL      1497.530958  1497.572083  0.492850  
BAL      1476.420665  1476.430696  0.490111  
BLB      1493.434841  1494.112711  0.399867  
...              ...          ...       ...  
WAS      1499.591579  1499.829463  0.428611  
WAT      1444.604816  1445.441439  0.348636  
WSA      1400.044113  1401.154166  0.517085  
WSB      1503.996104  1504.055381  0.471852  
WSC      1419.091395  1418.999671  0.557000  

[104 rows x 13 columns]

In [162]:
df_nba.groupby('team_id')['pts'].mean()

team_id
ANA    111.589744
AND     86.013889
ATL    101.920198
BAL    113.293303
BLB     79.698073
          ...    
WAS     96.551186
WAT     79.370968
WSA    118.395604
WSB    104.214859
WSC     76.931271
Name: pts, Length: 104, dtype: float64

In [163]:
df_nba.groupby('team_id')[['pts']].mean()

pts
team_id            
ANA      111.589744
AND       86.013889
ATL      101.920198
BAL      113.293303
BLB       79.698073
...             ...
WAS       96.551186
WAT       79.370968
WSA      118.395604
WSB      104.214859
WSC       76.931271

[104 rows x 1 columns]

In [164]:
# kita mau lihat juga rata2 skor lawannya berapa
df_nba.groupby('team_id')[['pts', 'opp_pts']].mean()

pts     opp_pts
team_id                        
ANA      111.589744  116.115385
AND       86.013889   82.555556
ATL      101.920198  102.259232
BAL      113.293303  113.711316
BLB       79.698073   83.349036
...             ...         ...
WAS       96.551186   98.915254
WAT       79.370968   84.903226
WSA      118.395604  119.736264
WSB      104.214859  105.227912
WSC       76.931271   74.965636

[104 rows x 2 columns]

In [165]:
# kita tambahkan groupby nya selain per tim, kita mau lihat per tahun juga
df_nba.groupby(['team_id', 'year_id'])[['pts', 'opp_pts']].mean()

pts     opp_pts
team_id year_id                        
ANA     1968     111.589744  116.115385
AND     1950      86.013889   82.555556
ATL     1969     110.215054  108.258065
        1970     117.307692  116.901099
        1971     113.459770  115.459770
...                     ...         ...
WSC     1947      72.803030   64.257576
        1948      73.612245   71.204082
        1949      81.126761   79.098592
        1950      76.728571   77.971429
        1951      81.257143   86.028571

[1487 rows x 2 columns]

In [175]:
# kita ambil beberapa tim tertentu saja, misalnya BOS dan LAL saja
df_boslal_2010 = df_nba[ ((df_nba['team_id'] == 'BOS') | (df_nba['team_id'] == 'LAL')) & 
                        (df_nba['year_id'] >= 2010)
                       ]

df_boslal_2010.groupby(['team_id', 'year_id'])[['pts', 'opp_pts']].mean()

pts     opp_pts
team_id year_id                        
BOS     2010      98.009434   94.556604
        2011      96.318681   91.351648
        2012      91.302326   89.139535
        2013      95.540230   96.114943
        2014      96.243902  100.707317
        2015     101.058140  101.337209
LAL     2010     101.561905   97.038095
        2011     100.576087   95.271739
        2012      96.794872   96.230769
        2013     101.418605  101.186047
        2014     102.951220  109.243902
        2015      98.451220  105.292683

# Manipulating Columns in DataFrame
- bisa buat kolom baru
- ubah nama kolom
- hapus kolom

In [178]:
df_nba.columns

Index(['gameorder', 'game_id', 'lg_id', '_iscopy', 'year_id', 'date_game',
       'seasongame', 'is_playoffs', 'team_id', 'fran_id', 'pts', 'elo_i',
       'elo_n', 'win_equiv', 'opp_id', 'opp_fran', 'opp_pts', 'opp_elo_i',
       'opp_elo_n', 'game_location', 'game_result', 'forecast', 'notes'],
      dtype='object')

In [179]:
# buat copy, biar data asli tidak terhapus
nba_copy = df_nba.copy()

In [185]:
# kita bisa setting custom untuk memperlihatkan semua kolom saat display dataframe
pd.set_option('display.max_columns', None)

In [186]:
# lalu kita editnya di copy nya aja
nba_copy.head(1)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947 1946-11-01           1   

   is_playoffs team_id  fran_id  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.2767   40.29483    NYK   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68     1300.0  1306.7233             H           L  0.640065   

  notes  
0   NaN

## buat kolom baru

In [187]:
# misalkan kita mau buat kolom untuk simpan data perbedaan skor
nba_copy['diff'] = nba_copy['pts'] - nba_copy['opp_pts']
nba_copy.head(1)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947 1946-11-01           1   

   is_playoffs team_id  fran_id  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.2767   40.29483    NYK   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68     1300.0  1306.7233             H           L  0.640065   

  notes  diff  
0   NaN    -2

## ubah nama kolom

In [191]:
nba_renamed = nba_copy.rename(columns={'fran_id':'team_name'}) # membuat df baru
nba_renamed.head(1)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947 1946-11-01           1   

   is_playoffs team_id team_name  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH   Huskies   66  1300.0  1293.2767   40.29483    NYK   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68     1300.0  1306.7233             H           L  0.640065   

  notes  diff  
0   NaN    -2

In [192]:
nba_copy.head(1) # rename tidak mengubah data awal

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947 1946-11-01           1   

   is_playoffs team_id  fran_id  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.2767   40.29483    NYK   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68     1300.0  1306.7233             H           L  0.640065   

  notes  diff  
0   NaN    -2

In [193]:
# ubah nama kolom sekaligus juga bisa
nba_renamed.columns

Index(['gameorder', 'game_id', 'lg_id', '_iscopy', 'year_id', 'date_game',
       'seasongame', 'is_playoffs', 'team_id', 'team_name', 'pts', 'elo_i',
       'elo_n', 'win_equiv', 'opp_id', 'opp_fran', 'opp_pts', 'opp_elo_i',
       'opp_elo_n', 'game_location', 'game_result', 'forecast', 'notes',
       'diff'],
      dtype='object')

In [197]:
nba_renamed.columns = ['gameorder', 'id', 'lg', '_copy', 'year', 'date',
       'season', '_playoffs', 'team_id', 'team_name', 'pts', 'elo_i',
       'elo_n', 'win_equiv', 'opp_id', 'opp_fran', 'opp_pts', 'opp_elo_i',
       'opp_elo_n', 'location', 'result', 'forecast', 'notes',
       'diff']

In [198]:
nba_renamed.head(1)

gameorder            id   lg  _copy  year       date  season  _playoffs  \
0          1  194611010TRH  NBA      0  1947 1946-11-01       1          0   

  team_id team_name  pts   elo_i      elo_n  win_equiv opp_id opp_fran  \
0     TRH   Huskies   66  1300.0  1293.2767   40.29483    NYK   Knicks   

   opp_pts  opp_elo_i  opp_elo_n location result  forecast notes  diff  
0       68     1300.0  1306.7233        H      L  0.640065   NaN    -2

## hapus kolom

In [199]:
nba_dropped = nba_renamed.drop(['elo_i', 'elo_n', 'win_equiv', 'opp_elo_i', 'opp_elo_n', 'forecast'], axis=1)
nba_dropped.head(1)

gameorder            id   lg  _copy  year       date  season  _playoffs  \
0          1  194611010TRH  NBA      0  1947 1946-11-01       1          0   

  team_id team_name  pts opp_id opp_fran  opp_pts location result notes  diff  
0     TRH   Huskies   66    NYK   Knicks       68        H      L   NaN    -2

### opsi inplace seperti menulis hasilnya langsung timpa data awal
```python
nba_renamed = nba_renamed.drop(['elo_i', 'elo_n', 'win_equiv', 'opp_elo_i', 'opp_elo_n', 'forecast'], axis=1)
nba_renamed.drop(['elo_i', 'elo_n', 'win_equiv', 'opp_elo_i', 'opp_elo_n', 'forecast'], axis=1, inplace=True)
```
efek kedua baris code ini sama

# ubah data type

In [201]:
nba_dropped['location'].value_counts()

H    63138
A    63138
N       38
Name: location, dtype: int64

untuk kolom 'location' data uniknya hanya ada 3, dan selalu akan hanya ada 3. kita bisa ubah tipe datanya dari "object" menjadi yang lebih spesifik yaitu "Categorical"

In [202]:
nba_dropped['location'] = pd.Categorical(nba_dropped['location'])

In [203]:
nba_dropped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 18 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   gameorder  126314 non-null  int64         
 1   id         126314 non-null  object        
 2   lg         126314 non-null  object        
 3   _copy      126314 non-null  int64         
 4   year       126314 non-null  int64         
 5   date       126314 non-null  datetime64[ns]
 6   season     126314 non-null  int64         
 7   _playoffs  126314 non-null  int64         
 8   team_id    126314 non-null  object        
 9   team_name  126314 non-null  object        
 10  pts        126314 non-null  int64         
 11  opp_id     126314 non-null  object        
 12  opp_fran   126314 non-null  object        
 13  opp_pts    126314 non-null  int64         
 14  location   126314 non-null  category      
 15  result     126314 non-null  object        
 16  notes      5424 non-

In [204]:
# keuntungan tipe data kategori
nba_dropped['location'].dtype

CategoricalDtype(categories=['A', 'H', 'N'], ordered=False)

In [206]:
# recap juga untuk tipe data tanggal bisa pakai pd.to_datetime()
nba_dropped['date'] = pd.to_datetime(nba_dropped['date'])

In [209]:
# keuntungan datetime, bisa baca per tahun, bulan, tanggal, hari
nba_dropped['tahun'] = nba_dropped['date'].dt.year
nba_dropped['bulan'] = nba_dropped['date'].dt.month
nba_dropped['tanggal'] = nba_dropped['date'].dt.day
nba_dropped.head(3)

gameorder            id   lg  _copy  year       date  season  _playoffs  \
0          1  194611010TRH  NBA      0  1947 1946-11-01       1          0   
1          1  194611010TRH  NBA      1  1947 1946-11-01       1          0   
2          2  194611020CHS  NBA      0  1947 1946-11-02       1          0   

  team_id team_name  pts opp_id opp_fran  opp_pts location result notes  diff  \
0     TRH   Huskies   66    NYK   Knicks       68        H      L   NaN    -2   
1     NYK    Knicks   68    TRH  Huskies       66        A      W   NaN     2   
2     CHS     Stags   63    NYK   Knicks       47        H      W   NaN    16   

   tahun  bulan  tanggal  
0   1946     11        1  
1   1946     11        1  
2   1946     11        2

In [210]:
# bisa juga lihat hari apa, senin selasa rabu..
nba_dropped['hari'] = nba_dropped['date'].dt.dayofweek
nba_dropped.head(3)

gameorder            id   lg  _copy  year       date  season  _playoffs  \
0          1  194611010TRH  NBA      0  1947 1946-11-01       1          0   
1          1  194611010TRH  NBA      1  1947 1946-11-01       1          0   
2          2  194611020CHS  NBA      0  1947 1946-11-02       1          0   

  team_id team_name  pts opp_id opp_fran  opp_pts location result notes  diff  \
0     TRH   Huskies   66    NYK   Knicks       68        H      L   NaN    -2   
1     NYK    Knicks   68    TRH  Huskies       66        A      W   NaN     2   
2     CHS     Stags   63    NYK   Knicks       47        H      W   NaN    16   

   tahun  bulan  tanggal  hari  
0   1946     11        1     4  
1   1946     11        1     4  
2   1946     11        2     5

untuk dayofweek, angkanya artinya:
- 0: Senin
- 1: Selasa
- 2: Rabu
- 3: Kamis
- 4: Jumat
- 5: Sabtu
- 6: Minggu

# Handling Missing Values
untuk data yang "bolong" atau kosong tidak ada isinya, kita bisa hapus dengan cepat menggunakan 
- .dropna() drop nya bisa untuk semua baris yang ada data kosong,
- .dropna(axis=1) bisa juga untuk semua kolom yang ada data kosong
- fillna() juga bisa kita gunakan untuk mengisi data2 kosong

In [211]:
nba_dropped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 22 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   gameorder  126314 non-null  int64         
 1   id         126314 non-null  object        
 2   lg         126314 non-null  object        
 3   _copy      126314 non-null  int64         
 4   year       126314 non-null  int64         
 5   date       126314 non-null  datetime64[ns]
 6   season     126314 non-null  int64         
 7   _playoffs  126314 non-null  int64         
 8   team_id    126314 non-null  object        
 9   team_name  126314 non-null  object        
 10  pts        126314 non-null  int64         
 11  opp_id     126314 non-null  object        
 12  opp_fran   126314 non-null  object        
 13  opp_pts    126314 non-null  int64         
 14  location   126314 non-null  category      
 15  result     126314 non-null  object        
 16  notes      5424 non-

In [215]:
# contoh hapus baris yang ada kosong
nba_dropna_baris = nba_dropped.dropna()
print(nba_dropna_baris.shape)
nba_dropna_baris.info()

(5424, 22)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 5424 entries, 930 to 124863
Data columns (total 22 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   gameorder  5424 non-null   int64         
 1   id         5424 non-null   object        
 2   lg         5424 non-null   object        
 3   _copy      5424 non-null   int64         
 4   year       5424 non-null   int64         
 5   date       5424 non-null   datetime64[ns]
 6   season     5424 non-null   int64         
 7   _playoffs  5424 non-null   int64         
 8   team_id    5424 non-null   object        
 9   team_name  5424 non-null   object        
 10  pts        5424 non-null   int64         
 11  opp_id     5424 non-null   object        
 12  opp_fran   5424 non-null   object        
 13  opp_pts    5424 non-null   int64         
 14  location   5424 non-null   category      
 15  result     5424 non-null   object        
 16  notes      5424 non-null   

In [216]:
# contoh hapus kolom yang ada kosong
nba_dropna_kolom = nba_dropped.dropna(axis=1)
print(nba_dropna_kolom.shape)
nba_dropna_kolom.info()

(126314, 21)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 21 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   gameorder  126314 non-null  int64         
 1   id         126314 non-null  object        
 2   lg         126314 non-null  object        
 3   _copy      126314 non-null  int64         
 4   year       126314 non-null  int64         
 5   date       126314 non-null  datetime64[ns]
 6   season     126314 non-null  int64         
 7   _playoffs  126314 non-null  int64         
 8   team_id    126314 non-null  object        
 9   team_name  126314 non-null  object        
 10  pts        126314 non-null  int64         
 11  opp_id     126314 non-null  object        
 12  opp_fran   126314 non-null  object        
 13  opp_pts    126314 non-null  int64         
 14  location   126314 non-null  category      
 15  result     126314 non-null  object        
 16  diff   

In [221]:
# pakai fillna untuk mengisi data2 yang kosong
nba_fillna = nba_dropped.copy()
nba_fillna['notes'] = nba_dropped['notes'].fillna('tidak ada notes')
nba_fillna.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 22 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   gameorder  126314 non-null  int64         
 1   id         126314 non-null  object        
 2   lg         126314 non-null  object        
 3   _copy      126314 non-null  int64         
 4   year       126314 non-null  int64         
 5   date       126314 non-null  datetime64[ns]
 6   season     126314 non-null  int64         
 7   _playoffs  126314 non-null  int64         
 8   team_id    126314 non-null  object        
 9   team_name  126314 non-null  object        
 10  pts        126314 non-null  int64         
 11  opp_id     126314 non-null  object        
 12  opp_fran   126314 non-null  object        
 13  opp_pts    126314 non-null  int64         
 14  location   126314 non-null  category      
 15  result     126314 non-null  object        
 16  notes      126314 no

In [222]:
nba_fillna.head()

gameorder            id   lg  _copy  year       date  season  _playoffs  \
0          1  194611010TRH  NBA      0  1947 1946-11-01       1          0   
1          1  194611010TRH  NBA      1  1947 1946-11-01       1          0   
2          2  194611020CHS  NBA      0  1947 1946-11-02       1          0   
3          2  194611020CHS  NBA      1  1947 1946-11-02       2          0   
4          3  194611020DTF  NBA      0  1947 1946-11-02       1          0   

  team_id team_name  pts opp_id  opp_fran  opp_pts location result  \
0     TRH   Huskies   66    NYK    Knicks       68        H      L   
1     NYK    Knicks   68    TRH   Huskies       66        A      W   
2     CHS     Stags   63    NYK    Knicks       47        H      W   
3     NYK    Knicks   47    CHS     Stags       63        A      L   
4     DTF   Falcons   33    WSC  Capitols       50        H      L   

             notes  diff  tahun  bulan  tanggal  hari  
0  tidak ada notes    -2   1946     11        1     4  
1  tidak ada notes     2   1946     11        1     4  
2  tidak ada notes    16   1946     11        2     5  
3  tidak ada notes   -16   1946     11        2     5  
4  tidak ada notes   -17   1946     11        2     5

In [223]:
nba_fillna['notes'].value_counts()

tidak ada notes           120890
at New York NY               440
at Philadelphia PA           256
at Norfolk VA                240
at Boston MA                 232
                           ...  
at Yardley PA                  2
at Providence, RI              2
at Bangor ME                   2
at Endicott NY                 2
at Wisconsin Rapids WI         2
Name: notes, Length: 232, dtype: int64

# Invalid Values
cek untuk data yang ada invalid nya

# Latihan
coba cari game milik tim Spurs dari tahun ke tahun di bulan 8, dan lihat rata2 skor nya berapa